File: core/ui/src/types/equipment.ts
import type { PageResponse } from "./grn";

export type EquipmentType =
  | "BALANCE"
  | "HPLC"
  | "GC"
  | "UV_SPECTROPHOTOMETER"
  | "IR_SPECTROPHOTOMETER"
  | "DISSOLUTION"
  | "PARTICLE_SIZE"
  | "KF_TITRATOR"
  | "PH_METER"
  | "TOC_ANALYZER"
  | "STABILITY_CHAMBER"
  | "REFRIGERATOR"
  | "AUTOCLAVE"
  | "LAB_COMPUTER"
  | "OTHER";

export type EquipmentStatus = "ACTIVE" | "UNDER_MAINTENANCE" | "RETIRED" | "PENDING_QUALIFICATION";

export type QualificationType = "IQ" | "OQ" | "PQ" | "REQUALIFICATION" | "CALIBRATION";

export type QualificationResult = "PASS" | "FAIL" | "CONDITIONAL_PASS" | "PENDING";

export type Equipment = {
  id: string;
  equipmentId: string;
  name: string;
  equipmentType: EquipmentType;
  manufacturer: string | null;
  model: string | null;
  serialNumber: string | null;
  location: string;
  status: EquipmentStatus;
  installationDate: string | null;
  lastQualificationDate: string | null;
  nextQualificationDue: string | null;
  lastCalibrationDate: string | null;
  nextCalibrationDue: string | null;
  calibrationIntervalMonths: number | null;
  responsibleAnalyst: string | null;
  calibrationOverdue: boolean;
  qualificationOverdue: boolean;
  daysUntilCalibrationDue: number | null;
  isActive: boolean;
  createdBy: string;
  createdAt: string;
  updatedBy: string | null;
  updatedAt: string | null;
};

export type QualificationRecord = {
  id: string;
  equipmentId: string;
  qualificationType: QualificationType;
  protocolReference: string;
  performedBy: string;
  performedAt: string;
  reviewedBy: string | null;
  reviewedAt: string | null;
  result: QualificationResult;
  deviationNoted: string | null;
  nextRequalificationDue: string | null;
  calibrationCertificateNumber: string | null;
  calibrationCertificateExpiry: string | null;
  eSignatureId: string | null;
  isActive: boolean;
  createdBy: string;
  createdAt: string;
};

export type CreateEquipmentRequest = {
  name: string;
  equipmentType: EquipmentType;
  location: string;
  manufacturer?: string;
  model?: string;
  serialNumber?: string;
  installationDate?: string;
  calibrationIntervalMonths?: number;
  responsibleAnalyst?: string;
};

export type UpdateEquipmentRequest = {
  name: string;
  location: string;
  manufacturer?: string;
  model?: string;
  serialNumber?: string;
  installationDate?: string;
  calibrationIntervalMonths?: number;
  responsibleAnalyst?: string;
  status: EquipmentStatus;
};

export type CreateQualificationRecordRequest = {
  qualificationType: QualificationType;
  protocolReference: string;
  performedBy: string;
  performedAt: string;
  reviewedBy?: string;
  reviewedAt?: string;
  result: QualificationResult;
  deviationNoted?: string;
  nextRequalificationDue?: string;
  calibrationCertificateNumber?: string;
  calibrationCertificateExpiry?: string;
  username?: string;
  password?: string;
  signatureMeaning?: string;
};

export type EquipmentSummary = {
  totalActive: number;
  pendingQualification: number;
  underMaintenance: number;
  calibrationDueSoon: number;
  qualificationDueSoon: number;
  calibrationOverdue: number;
  qualificationOverdue: number;
};

export type EquipmentPage = PageResponse<Equipment>;

File: core/src/main/java/com/batchsphere/core/lims/equipment/entity/Equipment.java
package com.batchsphere.core.lims.equipment.entity;

import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import jakarta.persistence.Id;
import jakarta.persistence.Table;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

import java.time.LocalDate;
import java.time.OffsetDateTime;
import java.util.UUID;

@Entity
@Table(name = "equipment")
@Getter
@Setter
@Builder
@NoArgsConstructor
@AllArgsConstructor
public class Equipment {

    @Id
    @Column(nullable = false, updatable = false)
    private UUID id;

    @Column(name = "equipment_id", nullable = false, unique = true, length = 50)
    private String equipmentId;

    @Column(nullable = false, length = 255)
    private String name;

    @Enumerated(EnumType.STRING)
    @Column(name = "equipment_type", nullable = false)
    private EquipmentType equipmentType;

    @Column(length = 255)
    private String manufacturer;

    @Column(length = 255)
    private String model;

    @Column(name = "serial_number", length = 100)
    private String serialNumber;

    @Column(nullable = false, length = 255)
    private String location;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private EquipmentStatus status;

    @Column(name = "installation_date")
    private LocalDate installationDate;

    @Column(name = "last_qualification_date")
    private LocalDate lastQualificationDate;

    @Column(name = "next_qualification_due")
    private LocalDate nextQualificationDue;

    @Column(name = "last_calibration_date")
    private LocalDate lastCalibrationDate;

    @Column(name = "next_calibration_due")
    private LocalDate nextCalibrationDue;

    @Column(name = "calibration_interval_months")
    private Integer calibrationIntervalMonths;

    @Column(name = "responsible_analyst", length = 100)
    private String responsibleAnalyst;

    @Column(name = "is_active", nullable = false)
    private Boolean isActive;

    @Column(name = "created_by", nullable = false, length = 100)
    private String createdBy;

    @Column(name = "created_at", nullable = false)
    private OffsetDateTime createdAt;

    @Column(name = "updated_by", length = 100)
    private String updatedBy;

    @Column(name = "updated_at")
    private OffsetDateTime updatedAt;
}
